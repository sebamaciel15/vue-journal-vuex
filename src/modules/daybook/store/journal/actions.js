import journalApi from '@/api/journalApi';
/* export const myAction = async({commit}) => {
    las ACTIONS pueden ser acciones asincronas que pueden llamar una mutacion
} */


export const loadEntries = async({commit}) => {
   const {data}= await journalApi.get('/entries.json')
    if(!data){
        commit('setEntries', [])
        return
    }

   const entries = []
   for(let id of Object.keys(data)){
       entries.push({
           id,
           ...data[id]
       })
   }
   commit('setEntries', entries)
}

export const updateEntry = async({commit}, entry) => { //entry debe de ser un parametro
    
    const {date, picture, text} = entry

    const dataToSave = {date, picture, text}

    const resp=  await journalApi.put(`/entries/${entry.id}.json`, dataToSave)

    dataToSave.id = entry.id

    commit('updateEntry', {...dataToSave})
    //extraer solo lo que se necesita //-id

    //awai journalApi.put( PATH .json, dataToSave)

    //commit de una mutation -> UpdateEntry
}

export const createEntry = async({commit}, entry) => {
    
    //dataToSave
    const {date, picture, text} = entry
    const dataToSave = {date, picture, text}

    const {data}=  await journalApi.post(`entries.json`, dataToSave)

    // data = {"name": "-MicDRRZDuEiUL-YJNdg"}
    dataToSave.id = data.name
    //commit  -> addEntry
    commit('addEntry', dataToSave)

    return data.name    
}

export const deleteEntry = async({commit}, id) => {
    await journalApi.delete(`/entries/${id}.json`)
    commit('deleteEntry', id)

    return id
}