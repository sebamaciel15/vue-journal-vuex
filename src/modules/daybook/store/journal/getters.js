
/* export const myGetter = (state) => {
    return state;
    los getters son para traer informacion del state y procesarla, bastante util para obtener entradas de una vista.
} */

export const getEntriesByTerm = ( state ) => ( term = '' ) => {

    if ( term.length === 0 ) return state.entries

    return state.entries.filter( entry => entry.text.toLowerCase().includes( term.toLocaleLowerCase() ) )
}

// id
export const getEntryById = ( state ) => ( id = '' ) => {

    const entry = state.entries.find( entry => entry.id === id )

    if ( !entry ) return

    return { ...entry } // TODO: prueben
}