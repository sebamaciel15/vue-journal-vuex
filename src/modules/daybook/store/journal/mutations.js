
/* export const myMutation = (state) => {
las mutaciones son sincronas y hacen la modificacion del state. el state es reactivo y cuando cambia notifica a todos los componentes que escuchen este cambio.
} */


export const setEntries = (state, entries ) => {
    state.entries = [...state.entries, ...entries]
    state.isLoading = false
    
}

export const updateEntry = (state, entry ) => { //entry actualizado
    
    const idx = state.entries.map(e => e.id ).indexOf(entry.id)
    state.entries[idx] = entry
    //state.entries = ... entry

}

export const addEntry = ( state, entry ) => {
    //state -> entries -> la nueva entrada debe ser la primera
    state.entries = [entry, ...state.entries]
}

export const deleteEntry= (state, id) => {
    //remover del state.entries borrar la entrada por ese id
    state.entries = state.entries.filter(entry => entry.id !== id)

}