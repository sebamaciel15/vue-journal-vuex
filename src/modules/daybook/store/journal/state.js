
export default () => ({
    isLoading: true,
    entries: [
        {
            id: new Date().getTime(),
            date: new Date().toDateString(),
            text: 'Ea proident ipsum pariatur consequat reprehenderit aute. Et labore enim id nisi anim nostrud duis. Nisi reprehenderit dolor occaecat consectetur consequat Lorem ex nostrud occaecat. Minim aliqua magna enim laboris do.',
            picture: null,
        },
        {
            id: new Date().getTime() + 1000,
            date: new Date().toDateString(),
            text: 'In sunt consequat veniam do cillum deserunt id in veniam eiusmod magna commodo sint. Ad esse ut ea dolor fugiat cillum nulla ad non consectetur irure anim. Non amet aliquip tempor fugiat. Est eu voluptate id aliquip mollit culpa aute ullamco velit eiusmod reprehenderit id. Ipsum eu nulla consequat pariatur culpa.',
            picture: null,
        },
        {
            id: new Date().getTime() + 2000,
            date: new Date().toDateString(),
            text: 'Ad incididunt ullamco consectetur consequat quis ex nulla do enim occaecat anim aliquip est. Nulla ea cupidatat magna irure ex anim veniam veniam fugiat aliqua ex et aliqua velit. Nulla incididunt nulla enim cupidatat non. Irure pariatur laboris voluptate anim fugiat irure pariatur dolore officia do consequat reprehenderit veniam ea. Lorem dolore voluptate deserunt anim eiusmod officia irure sint mollit enim ea nulla.',
            picture: null,
        },
    ],
})