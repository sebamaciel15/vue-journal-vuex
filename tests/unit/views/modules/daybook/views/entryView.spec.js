import {shallowMount} from '@vue/test-utils'
import { createStore } from 'vuex'
import journal from '@/modules/daybook/store/journal'
import {journalState} from '../../../mock-data/test-journal-state'
import EntryView from '@/modules/daybook/views/EntryView.vue';
import Swal from 'sweetalert2'

const createVuexStore= (initialState) => 
        createStore({
            modules: {
                journal: {
                ...journal,
                state: {...initialState}
                }
            }
        })
jest.mock('sweetalert2', () => ({
    fire: jest.fn(),
    showLoading: jest.fn(),
    close: jest.fn()
}))
describe('Pruebas en el EntryView', () => {
    const store= createVuexStore(journalState)
    store.dispatch = jest.fn()
    const mockRouter = {
        push: jest.fn()
    }
    let wrapper
    beforeEach(() => {
        jest.clearAllMocks()
        wrapper = shallowMount(EntryView, {
            props: {
                id: '-MimEuhIuzcMLq3uy_8T'
            },
            global: {
                mocks: {
                     $router: mockRouter
                },
                plugins: [store]
            }
        })
    })    
    test('debe sacar al usuario porque el ID no existe', () => {
        const wrapper = shallowMount(EntryView, {
            props: {
                id: 'Este ID no existe en el STORE'
            },
            global: {
                mocks: {
                     $router: mockRouter
                },
                plugins: [store]
            }
        })
        expect(mockRouter.push).toHaveBeenCalledWith({name: 'no-entry'})
    })    
    test('debe mostrar la entrada correctamente', () => {
        expect(wrapper.html()).toMatchSnapshot()
        expect(mockRouter.push).not.toHaveBeenCalled()
    })  
    test('debe borrar entrada y salir', (done) => {
        Swal.fire.mockReturnValueOnce(Promise.resolve({isConfirmed:true}))
        wrapper.find('.btn-danger').trigger('click')
        expect(Swal.fire).toHaveBeenCalledWith({
            title: 'Â¿Estas seguro?',
            text: 'Una vez borrado, no se puede recuperar.',
            showDenyButton: true,
            confirmButtonText: 'Si, estoy seguro'
        })
        setTimeout( () => {
            expect(store.dispatch).toHaveBeenCalledWith('journal/deleteEntry', '-MimEuhIuzcMLq3uy_8T')
            expect(mockRouter.push).toHaveBeenCalled()    
            done()        
        }, 1)
    })
      
})
