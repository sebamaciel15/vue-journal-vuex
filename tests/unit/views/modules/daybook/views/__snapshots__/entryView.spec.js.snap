// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el EntryView debe mostrar la entrada correctamente 1`] = `
<div class="entry-title d-flex justifi-content-between p-2">
  <div><span class="text-success fs-3 fw-bold">6</span><span class="mx-1 fs-3">Septiembre</span><span class="mx-2 fs-4 fw-light">2021, Jueves </span></div>
  <div><input type="file" accept="image/png, image/jpeg" style="display: none;"><button class="btn btn-danger mx-2"> Borrar <i class="fa fa-trash-alt"></i></button><button class="btn btn-primary"> Subir foto <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="Â¿Que sucedio hoy?"></textarea></div><img src="https://res.cloudinary.com/dg4dflmtg/image/upload/v1630783121/wezqsrgrvssaiziswxcx.png" alt="entry picture" class="img-thumbnail">
<!--v-if-->
<fab-stub icon="fa-save"></fab-stub>
`;
